compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Subst.hs

```diff
@@ -83,13 +101,15 @@ instance SubstTm Expr where
     ETmApp e1 e2 -> ETmApp (substituteTm s e1) (substituteTm s e2)
     ETyApp e t -> ETyApp (substituteTm s e) t
     ETmLam (x,t) e -> if x `elem` substDom s
-      then ETmLam (x,t) e
+      then ETmLam (x,t) (substituteTm (removeSubst s x) e)
       else ETmLam (x,t) (substituteTm s e)
     ETyLam (a,k) e -> ETyLam (a,k) (substituteTm s e)
     ECase e cs -> ECase (substituteTm s e)
       $ map (\CaseAlternative{..} -> CaseAlternative altPattern (substituteTm s altExpr)) cs
-    ELet Binding{..} e -> ELet (Binding bindingBinder $ substituteTm s bindingBound)
-      (substituteTm s e)
+    ELet (Binding (var,ty) e1) e2 -> if var `elem` substDom s
+      then let s' = removeSubst s var
+           in ELet (Binding (var,ty) (substituteTm s e1)) (substituteTm s' e2)
+      else ELet (Binding (var,ty) (substituteTm s e1)) (substituteTm s e2)
```

Comment:
You should be able to unconditionally remove the `var` from the substitution. If it's not in the domain, the substitution won't change.
```suggestion
    ELet (Binding (var,ty) e1) e2 ->
      let s' = removeSubst s var
      in ELet (Binding (var,ty) (substituteTm s e1)) (substituteTm s' e2)
```
The same applies to `ETmLam` above.